src/System.Diagnostics.StackTrace.Symbols/src/System/Diagnostics/StackTrace/Symbols.cs

```diff
@@ -0,0 +1,222 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using System.IO;
+using System.Reflection.Metadata;
+using System.Reflection.Metadata.Ecma335;
+using System.Reflection.PortableExecutable;
+
+namespace System.Diagnostics.StackTrace
+{
+    public static class Symbols
+    {
+        /// <summary>
+        /// Returns the source file and line number information for the method.
+        /// </summary>
+        /// <param name="assemblyPath">file path of the assembly or null</param>
+        /// <param name="inMemorySymbols">in memory PDB address or zero</param>
+        /// <param name="inMemorySymbolsSize">in memory PDB size</param>
+        /// <param name="methodToken">method token</param>
+        /// <param name="ilOffset">il offset of the stack frame</param>
+        /// <param name="sourceFile">source file return</param>
+        /// <param name="sourceLine">line number return</param>
+        /// <param name="sourceColumn">column return</param>
+        public static void GetSourceLineInfo(string assemblyPath, IntPtr inMemorySymbols, int inMemorySymbolsSize, 
+            int methodToken, int ilOffset, out string sourceFile, out int sourceLine, out int sourceColumn)
+        {
+            sourceFile = null;
+            sourceLine = 0;
+            sourceColumn = 0;
+
+            MetadataReader reader;
+            using (GetReader(assemblyPath, inMemorySymbols, inMemorySymbolsSize, out reader)) {
+                if (reader != null) {
+                    Handle handle = MetadataTokens.Handle(methodToken);
+
+                    if (handle.Kind == HandleKind.MethodDefinition) {
+                        MethodDebugInformationHandle methodDebugHandle = ((MethodDefinitionHandle)handle).ToDebugInformationHandle();
+                        MethodDebugInformation methodInfo = reader.GetMethodDebugInformation(methodDebugHandle);
+
+                        if (!methodInfo.SequencePointsBlob.IsNil) {
+                            try {
+                                SequencePointCollection sequencePoints = methodInfo.GetSequencePoints();
+
+                                int sequencePointCount = 0;
+                                foreach (SequencePoint sequence in sequencePoints) {
```

Comment:
I think 47-88 could be rewritten more simply like this?

    SequencePoint? bestPointSoFar = null;
    foreach(SequencePoint p in sequencePoints)
    {
        if(p.Offset > ilOffset) break;
        if(p.StartLine != SequencePoint.HiddenLine)  bestPointSoFar = p;
    }